chia/full_node/full_node_api.py

```diff
@@ -1720,5 +1712,201 @@ def get_fee_estimates(est: FeeEstimatorInterface, req_times: List[uint64]) -> Li
         msg = make_msg(ProtocolMessageTypes.respond_fee_estimates, response)
         return msg
 
+    @api_request(
+        peer_required=True,
+        reply_types=[ProtocolMessageTypes.respond_add_puzzle_subscriptions],
+    )
+    async def request_add_puzzle_subscriptions(
+        self, request: wallet_protocol.RequestAddPuzzleSubscriptions, peer: WSChiaConnection
+    ) -> Message:
+        peer_id = peer.peer_node_id
+        max_subscriptions = self.max_subscriptions(peer)
+        subs = self.full_node.subscriptions
+
+        peak_height = self.full_node.blockchain.get_peak_height()
+        assert peak_height is not None
+
+        header_hash = self.full_node.blockchain.height_to_hash(peak_height)
+        assert header_hash is not None
+
+        new_subscriptions = set(request.puzzle_hashes) - subs.puzzle_subscriptions(peer_id)
+        added = subs.add_puzzle_subscriptions(peer_id, list(new_subscriptions), max_subscriptions)
+
+        response = wallet_protocol.RespondAddPuzzleSubscriptions(list(added), peak_height, header_hash)
+        msg = make_msg(ProtocolMessageTypes.respond_add_puzzle_subscriptions, response)
+        return msg
+
+    @api_request(
+        peer_required=True,
+        reply_types=[ProtocolMessageTypes.respond_add_coin_subscriptions],
+    )
+    async def request_add_coin_subscriptions(
+        self, request: wallet_protocol.RequestAddCoinSubscriptions, peer: WSChiaConnection
+    ) -> Message:
+        peer_id = peer.peer_node_id
+        max_subscriptions = self.max_subscriptions(peer)
+        subs = self.full_node.subscriptions
+
+        peak_height = self.full_node.blockchain.get_peak_height()
+        assert peak_height is not None
+
+        header_hash = self.full_node.blockchain.height_to_hash(peak_height)
+        assert header_hash is not None
+
+        new_subscriptions = set(request.coin_ids) - subs.coin_subscriptions(peer_id)
+        added = subs.add_coin_subscriptions(peer_id, list(new_subscriptions), max_subscriptions)
+
+        response = wallet_protocol.RespondAddCoinSubscriptions(list(added), peak_height, header_hash)
+        msg = make_msg(ProtocolMessageTypes.respond_add_coin_subscriptions, response)
+        return msg
+
+    @api_request(
+        peer_required=True,
+        reply_types=[ProtocolMessageTypes.respond_remove_puzzle_subscriptions],
+    )
+    async def request_remove_puzzle_subscriptions(
+        self, request: wallet_protocol.RequestRemovePuzzleSubscriptions, peer: WSChiaConnection
+    ) -> Message:
+        peer_id = peer.peer_node_id
+        subs = self.full_node.subscriptions
+
+        if request.puzzle_hashes is None:
+            removed = list(subs.puzzle_subscriptions(peer_id))
+            subs.remove_puzzle_subscriptions(peer_id, removed)
```

Comment:
I would expect it to be a little bit more efficient to just tell `subs` to remove all the subscriptions for a peer, rather than asking for them and removing them one at a time. but maybe it's not worth the hassle right now.

I suppose we're returning the ones we removed anyway.

Reply:
Seems reasonable, I think there's already a method to remove all puzzle subscriptions, and if not I can add it

Reply:
Implemented